common/kotlinx-coroutines-core-common/src/Job.kt

```diff
@@ -97,22 +103,32 @@ public interface Job : CoroutineContext.Element {
     // ------------ state query ------------
 
     /**
-     * Returns `true` when this job is active -- it was already started and has not completed or cancelled yet.
+     * Returns `true` when this job is active -- it was already started and has not completed or failed yet.
      * The job that is waiting for its [children] to complete is still considered to be active if it
-     * was not cancelled.
+     * has not failed and was not cancelled.
      */
     public val isActive: Boolean
 
     /**
-     * Returns `true` when this job has completed for any reason. A job that was cancelled and has
-     * finished its execution is also considered complete. Job becomes complete only after
+     * Returns `true` when this job has completed for any reason. A job that has failed or cancelled
+     * and has finished its execution is also considered complete. Job becomes complete only after
      * all its [children] complete.
      */
     public val isCompleted: Boolean
 
     /**
-     * Returns `true` if this job was [cancelled][cancel]. In the general case, it does not imply that the
-     * job has already [completed][isCompleted] (it may still be cancelling whatever it was doing).
+     * Returns `true` if the failure was detected while running this job and it cannot complete normally.
+     * In the general case, it does not imply that the
+     * job has already [completed][isCompleted], because it may still be finishing whatever it was doing and
+     * waiting for its [children] to complete.
+     */
+    public val isFailed: Boolean
+
+    /**
+     * Returns `true` if this job was [cancelled][cancel]. It implies that [isFailed] returns `true`, too.
+     * In the general case, it does not imply that the
+     * job has already [completed][isCompleted], because it may still be finishing whatever it was doing and
+     * failing for its [children] to complete.
      */
     public val isCancelled: Boolean
```

Comment:
That is indeed a serious problem that arises from the fact that we try to maintain failed / cancelled distinction. Indeed, when child has a parent and fail, then it fails parent, which fails all the children including _this one_. I do not immediately see a nice solution to it.

Reply:
The solution that immediately comes to mind is to _never cancel on failure_. Instead of cancelling children make them _failed_.  That is, make a coroutine _cancelled_ **if and only if** a `cancel` was explicitly invoked.